cpp/src/parquet/column_reader.cc

```diff
@@ -1957,6 +1970,139 @@ class ByteArrayChunkedRecordReader : public TypedRecordReader<ByteArrayType>,
   typename EncodingTraits<ByteArrayType>::Accumulator accumulator_;
 };
 
+class ByteArrayChunkedOptRecordReader : public TypedRecordReader<ByteArrayType>,
+                                        virtual public BinaryRecordReader {
+ public:
+  ByteArrayChunkedOptRecordReader(const ColumnDescriptor* descr, LevelInfo leaf_info,
+                                  ::arrow::MemoryPool* pool)
+      : TypedRecordReader<ByteArrayType>(descr, leaf_info, pool) {
+    DCHECK_EQ(descr_->physical_type(), Type::BYTE_ARRAY);
+    accumulator_.builder.reset(new ::arrow::BinaryBuilder(pool));
+    values_ = AllocateBuffer(pool);
+    offset_ = AllocateBuffer(pool);
+  }
+
+  ::arrow::ArrayVector GetBuilderChunks() override {
+    if (uses_opt_) {
+      std::vector<std::shared_ptr<Buffer>> buffers = {ReleaseIsValid(), ReleaseOffsets(),
+                                                      ReleaseValues()};
+      auto data = std::make_shared<::arrow::ArrayData>(
+          ::arrow::binary(), values_written(), buffers, null_count());
+
+      auto chunks = ::arrow::ArrayVector({::arrow::MakeArray(data)});
+      return chunks;
+    } else {
+      ::arrow::ArrayVector result = accumulator_.chunks;
+      if (result.size() == 0 || accumulator_.builder->length() > 0) {
+        std::shared_ptr<::arrow::Array> last_chunk;
+        PARQUET_THROW_NOT_OK(accumulator_.builder->Finish(&last_chunk));
+        result.push_back(std::move(last_chunk));
+      }
+      accumulator_.chunks = {};
+      return result;
```

Comment:
The optimized RecordReader implementation is ByteArrayChunkedRecordReader

 https://github.com/zhixingheyi-tian/arrow/blob/8d611c673d3ed23abfa32f50da3cc4c4e4feb807/cpp/src/parquet/column_reader.cc#L2203-L2206

And is just for  Binary/String/LargeBinary/LargeString types.

Reply:
This probably means it's not good to create a new class for this optimized reader.

That said, I don't have deep knowlege of parquet code and lack bandwidth recently to investigate, someone else might comment.

Reply:
I agree, would shouldn't create another class. I don't see any reason this can't be used for the Decimal case.

Reply:
I've found locally that if I merge the implementations, the unit tests pass. Could you please merge them in the PR?

Reply:
@wjones127 
Do you merge the two class: ByteArrayChunkedRecordReader and ByteArrayChunkedOptRecordReader, ,  all unit tests passed?

Reply:
Yes, those are the two classes I merged.

Reply:
Here is the changeset that allows the `parquet-arrow-test` and `parquet-arrow-internals-test` to pass. https://github.com/zhixingheyi-tian/arrow/pull/2

Could you incorporate those changes?

Reply:
+1 for not adding a separate class. This would be difficult to maintain if more optimization will be added. It would be better if an option can be added so that user can manually turn it off when something goes wrong with the new feature.